allennlp/modules/elmo.py

```diff
@@ -455,11 +469,26 @@ class _ElmoBiLm(torch.nn.Module):
     def __init__(self,
                  options_file: str,
                  weight_file: str,
-                 requires_grad: bool = False) -> None:
+                 requires_grad: bool = False,
+                 vocab_to_cache: List[str] = None) -> None:
         super(_ElmoBiLm, self).__init__()
 
         self._token_embedder = _ElmoCharacterEncoder(options_file, weight_file, requires_grad=requires_grad)
 
+        self._requires_grad = requires_grad
+        if requires_grad and vocab_to_cache:
+            logging.warning("You are fine tuning ELMo and caching char CNN word vectors. "
+                            "This behaviour is not guaranteed to be well defined, particularly. "
+                            "if not all of your inputs will occur in the vocabulary cache.")
+
+        self._word_embedding = None
+        if vocab_to_cache:
+            logging.info("Caching character cnn layers for words in vocabulary.")
+            # This sets 3 attributes, _word_embedding, _bos_embedding and _eos_embedding.
+            # They are set in the method so it can be accessed from outside the
```

Comment:
`s/so it can/so they can/`.  Also, shouldn't you have to set `_bos_embedding` and `_eos_embedding` as attributes on `self` here?  I'm surprised pylint isn't complaining about this.  And type annotations on these initial `None` assignments would be nice; I'm also surprised mypy isn't complaining about that.

Reply:
So this is annoying - If I import the Embedding, mypy complains because I have to do it before the module is imported, and then complains again because None isn't callable. I think the others don't complain because the type annotation on `add_sentence_boundaries` is `Any`, so it doesn't matter what they are.

Reply:
Hmm, interesting, ok.  The point of the annotations is mostly to aid readability; if it's too much work (or impossible) to get mypy to be happy with it, then a simple comment explaining what `_word_embedding` holds would be good enough.